hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -501,7 +501,17 @@ private Boolean destroyCurrentClusterConnectionAndTryNextCluster(CandidateCluste
         logger.info("Trying to connect to next cluster: " + nextContext.getClusterName());
 
         if (doConnectToCandidateCluster(nextContext, true)) {
-            client.waitForInitialMembershipEvents();
+            if (!failoverConfigProvided) {
+                client.waitForInitialMembershipEvents();
+            } else {
+                // If the client is a failover client, the expected behaviour in case it can't get the initial membership event
+                // is that it should continue trying other clusters.
+                try {
+                    client.waitForInitialMembershipEvents();
+                } catch (IllegalStateException e) {
+                    return false;
```

Comment:
you return here but when are you closing the opened connection to the cluster?

Reply:
it closes it automatically. check code around destroy current context